File: Course_1-Algorithmic_Toolbox/Week-5/Excercise_Challenges/5_longest_common_subsequence_of_three_sequences/lcs3.py
import numpy as np

def lcs3(a, b, c):
    n, m, l = len(a), len(b), len(c)
    dp = np.zeros((n + 1, m + 1, l + 1))

    for i in range(n + 1):
        for j in range(m + 1):
            for k in range(l + 1):
                if i == 0 or j == 0 or k == 0:
                    dp[i][j][k] = 0
                elif a[i - 1] == b[j - 1] and a[i - 1] == c[k - 1]:
                    dp[i][j][k] = 1 + dp[i - 1][j - 1][k - 1]
                else:
                    dp[i][j][k] = max(dp[i - 1][j][k],
                                      dp[i][j - 1][k],
                                      dp[i][j][k - 1])
    
    return int(dp[n][m][l])

if __name__ == '__main__':
    n = int(input())
    a = list(map(int, input().split()))

    n = int(input())
    b = list(map(int, input().split()))

    n = int(input())
    c = list(map(int, input().split()))
    print(lcs3(a, b, c))

File: Course_1-Algorithmic_Toolbox/Week-5/Excercise_Challenges/5_longest_common_subsequence_of_three_sequences/lcs3.cpp
#include <iostream>
#include <vector>
using namespace std;

int max3(int a, int b, int c)
{
    return max(max(a, b), c);
}

int lcs3(vector<int> &a, vector<int> &b, vector<int> &c)
{
    int n = a.size(), m = b.size(), l = c.size();
    int dp[n + 1][m + 1][l + 1];


    for(int i = 0; i < n + 1; ++i){
        for(int j = 0; j < m + 1; ++j){
            for(int k = 0; k < l + 1; ++k){
                if(i == 0 || j == 0 || k == 0)
                    dp[i][j][k] = 0;
                else if(a[i - 1] == b[j - 1] && a[i - 1] == c[k - 1])
                    dp[i][j][k] = 1 + dp[i - 1][j - 1][k - 1];
                else
                    dp[i][j][k] = max3(dp[i - 1][j][k], dp[i][j - 1][k], dp[i][j][k - 1]);
            }
        }
    }

    return dp[n][m][l];
}

int main(void)
{
    int an;
    cin >> an;
    vector<int> a(an);
    for(int i = 0; i < an; i++){
        cin >> a[i];
    }
  
    int bn;
    cin >> bn;
    vector<int> b(bn);
    for(int i = 0; i < bn; i++){
        cin >> b[i];
    }

    int cn;
    cin >> cn;
    vector<int> c(cn);
    for(int i = 0; i < cn; i++){
        cin >> c[i];
    }

    cout << lcs3(a, b, c) << "\n";

    return 0;
}
